File: projetos-fenix-para-estudo/t2ti-erp-fenix-banco-agencia-pessoa-completos/fontes/backend/c-sharp/T2TiERPFenix/T2TiERPFenix/Controllers/Cadastros/BancoController.cs
using Microsoft.AspNetCore.Mvc;
using System;
using T2TiERPFenix.Models;
using T2TiERPFenix.Repository;
using Microsoft.AspNetCore.Http;
using System.Collections.Generic;

namespace T2TiERPFenix.Controllers
{
    [Route("banco")]
    [Produces("application/json")]
    public class BancoController : Controller
    {
        private IRepositoryWrapper _repository;

        public BancoController(IRepositoryWrapper repository)
        {
            _repository = repository;
        }

        [HttpGet]
        public IActionResult ConsultarListaBanco([FromQuery]string filter)
        {
            try
            {
                IEnumerable<Banco> lista;
                if (filter == null)
                {
                    lista = _repository.Banco.ConsultarLista();
                } else
                {
                    // define o filtro
                    Filtro filtro = new Filtro(filter);
                    lista = _repository.Banco.ConsultarListaFiltro(filtro);
                }
                return Ok(lista);
            }
            catch (Exception ex)
            {
                return StatusCode(500, new RetornoJsonErro(500, "Erro no Servidor [Consultar Lista Banco]", ex));
            }
        }

        [HttpGet("{id}", Name = "ConsultarObjetoBanco")]
        public IActionResult ConsultarObjetoBanco(int id)
        {
            try
            {
                var objeto = _repository.Banco.ConsultarObjeto(id);

                if (objeto == null)
                {
                    return StatusCode(404, new RetornoJsonErro(404, "Registro não localizado [Consultar Objeto Banco]", null));
                }
                else
                {
                    return Ok(objeto);
                }
            }
            catch (Exception ex)
            {
                return StatusCode(500, new RetornoJsonErro(500, "Erro no Servidor [Consultar Objeto Banco]", ex));
            }
        }

        [HttpPost]
        public IActionResult InserirBanco([FromBody]Banco objJson)
        {
            try
            {
                if (!ModelState.IsValid)
                {
                    return StatusCode(400, new RetornoJsonErro(400, "Objeto inválido [Inserir Banco]", null));
                }

                _repository.Banco.Inserir(objJson);

                return CreatedAtRoute("ConsultarObjetoBanco", new { id = objJson.Id }, objJson);
            }
            catch (Exception ex)
            {
                return StatusCode(500, new RetornoJsonErro(500, "Erro no Servidor [Inserir Banco]", ex));
            }
        }

        [HttpPut("{id}")]
        public IActionResult AlterarBanco([FromBody]Banco objJson, int id)
        {
            try
            {
                if (!ModelState.IsValid)
                {
                    return StatusCode(400, new RetornoJsonErro(400, "Objeto inválido [Alterar Banco]", null));
                }

                if (objJson.Id != id)
                {
                    return StatusCode(400, new RetornoJsonErro(400, "Objeto inválido [Alterar Banco] - ID do objeto difere do ID da URL.", null));
                }

                var objBanco = _repository.Banco.ConsultarObjeto(objJson.Id);

                if (objBanco == null)
                {
                    return StatusCode(400, new RetornoJsonErro(400, "Objeto com ID inválido [Alterar Banco]", null));
                }

                _repository.Banco.Alterar(objBanco, objJson);

                return ConsultarObjetoBanco(id);
            }
            catch (Exception ex)
            {
                return StatusCode(500, new RetornoJsonErro(500, "Erro no Servidor [Alterar Banco]", ex));
            }
        }

        [HttpDelete("{id}")]
        public IActionResult ExcluirBanco(int id)
        {
            try
            {
                var objeto = _repository.Banco.ConsultarObjeto(id);

                _repository.Banco.Excluir(objeto);

                return Ok();
            }
            catch (Exception ex)
            {
                return StatusCode(500, new RetornoJsonErro(500, "Erro no Servidor [Excluir Banco]", ex));
            }
        }

    }
}

File: projetos-fenix-para-estudo/t2ti-erp-fenix-servidores-provider-inicio/fontes/backend/delphi/dmvc/Controller/BancoController.pas
unit BancoController;

interface

uses mvcframework, mvcframework.Commons,
  System.SysUtils,
  MVCFramework.SystemJSONUtils;

type

  [MVCDoc('CRUD Banco')]
  [MVCPath('/banco')]
  TBancoController = class(TMVCController)
  public
    [MVCDoc('Retorna uma lista de objetos')]
    [MVCPath('/($filtro)')]
    [MVCHTTPMethod([httpGET])]
    procedure ConsultarLista(Context: TWebContext);

//    [MVCDoc('Retorna uma lista de objetos com base no Filtro')]
//    [MVCPath('/($campo)/($valor)')]
//    [MVCHTTPMethod([httpGET])]
//    procedure ConsultarListaFiltroValor(campo: string; valor: string);

    [MVCDoc('Retorna um objeto com base no ID')]
    [MVCPath('/($id)')]
    [MVCHTTPMethod([httpGET])]
    procedure ConsultarObjeto(id: Integer);

    [MVCDoc('Insere um novo objeto')]
    [MVCPath]
    [MVCHTTPMethod([httpPOST])]
    procedure Inserir(Context: TWebContext);

    [MVCDoc('Altera um objeto com base no ID')]
    [MVCPath('/($id)')]
    [MVCHTTPMethod([httpPUT])]
    procedure Alterar(id: Integer);

    [MVCDoc('Exclui um objeto com base no ID')]
    [MVCPath('/($id)')]
    [MVCHTTPMethod([httpDelete])]
    procedure Excluir(id: Integer);
  end;

implementation

{ TBancoController }

uses BancoService, Banco, Commons, Filtro;

procedure TBancoController.ConsultarLista(Context: TWebContext);
var
  FiltroUrl, FilterWhere: string;
  FiltroObj: TFiltro;
begin
  FiltroUrl := Context.Request.Params['filtro'];
  if FiltroUrl <> '' then
  begin
    ConsultarObjeto(StrToInt(FiltroUrl));
    exit;
  end;

  filterWhere := Context.Request.Params['filter'];
  try
    if FilterWhere = '' then
    begin
      Render<TBanco>(TBancoService.ConsultarLista);
    end
    else begin
      // define o objeto filtro
      FiltroObj := TFiltro.Create(FilterWhere);
      Render<TBanco>(TBancoService.ConsultarListaFiltroValor(FiltroObj.Campo, FiltroObj.Valor));
    end;
  except
    on E: EServiceException do
    begin
      raise EMVCException.Create
        ('Erro no Servidor [Consultar Lista Banco] - Exceção: ' + E.Message,
        E.StackTrace, 0, 500);
    end
    else
      raise;
  end;
end;

//procedure TBancoController.ConsultarListaFiltroValor(campo, valor: string);
//begin
//  try
//    Render<TBanco>(TBancoService.ConsultarListaFiltroValor(campo, valor));
//  except
//    on E: EServiceException do
//    begin
//      raise EMVCException.Create
//        ('Erro no Servidor [Consultar Lista Filtro Valor Banco] - Exceção: ' + E.Message,
//        E.StackTrace, 0, 500);
//    end
//    else
//      raise;
//  end;
//end;

procedure TBancoController.ConsultarObjeto(id: Integer);
var
  Banco: TBanco;
begin
  try
    Banco := TBancoService.ConsultarObjeto(id);

    if Assigned(Banco) then
      Render(Banco)
    else
      raise EMVCException.Create
        ('Registro não localizado [Consultar Objeto Banco]', '', 0, 404);
  except
    on E: EServiceException do
    begin
      raise EMVCException.Create
        ('Erro no Servidor [Consultar Objeto Banco] - Exceção: ' + E.Message,
        E.StackTrace, 0, 500);
    end
    else
      raise;
  end;
end;

procedure TBancoController.Inserir(Context: TWebContext);
var
  Banco: TBanco;
begin
  try
    Banco := Context.Request.BodyAs<TBanco>;
  except
    on E: EServiceException do
    begin
      raise EMVCException.Create('Objeto inválido [Inserir Banco] - Exceção: ' +
        E.Message, E.StackTrace, 0, 400);
    end
    else
      raise;
  end;

  try
    TBancoService.Inserir(Banco);
    Render(200, 'Objeto inserido com sucesso.');
  finally
    FreeAndNil(Banco);
  end;
end;

procedure TBancoController.Alterar(id: Integer);
var
  Banco, BancoDB: TBanco;
begin
  try
    Banco := Context.Request.BodyAs<TBanco>;
  except
    on E: EServiceException do
    begin
      raise EMVCException.Create('Objeto inválido [Alterar Banco] - Exceção: ' +
        E.Message, E.StackTrace, 0, 400);
    end
    else
      raise;
  end;

  if Banco.Id <> id then
    raise EMVCException.Create('Objeto inválido [Alterar Banco] - ID do objeto difere do ID da URL.',
      '', 0, 400);

  BancoDB := TBancoService.ConsultarObjeto(Banco.id);

  if not Assigned(BancoDB) then
    raise EMVCException.Create('Objeto com ID inválido [Alterar Banco]',
      '', 0, 400);

  try
    TBancoService.Alterar(Banco);
    Render(200, 'Objeto alterado com sucesso.');
  finally
    FreeAndNil(Banco);
    FreeAndNil(BancoDB);
  end;
end;

procedure TBancoController.Excluir(id: Integer);
var
  Banco: TBanco;
begin
  Banco := TBancoService.ConsultarObjeto(id);

  if not Assigned(Banco) then
    raise EMVCException.Create('Objeto com ID inválido [Excluir Banco]',
      '', 0, 400);

  try
    TBancoService.Excluir(Banco);
    Render(200, 'Objeto excluído com sucesso.');
  finally
    FreeAndNil(Banco)
  end;
end;

end.
